swag/bank.py

```diff
@@ -301,6 +309,216 @@ def set_timezone(self, user, timezone):
 
         return lock_date
 
+    ## Ajout des fonctions des cagnottes
+
+    def create_cagnotte(self, cagnotte_name: str, currency: Currency, creator_id: int):
+        self.swagdb.add_cagnotte(cagnotte_name, currency, creator_id)
+        self.transactional_save()
+        return self.swagdb.cagnotte_number() - 1  # return the idx of the Cagnotte
+
+    def get_active_cagnotte(self, cagnotte_idx) -> Cagnotte:
+        if cagnotte_idx in self.swagdb.get_active_cagnotte_ids():
```

Comment:
```python
    def get_active_cagnotte(self, cagnotte_id) -> Cagnotte:
        cagnotte = self.swagdb.get_cagnotte(cagnotte_id)
        if cagnotte.is_active:
            return cagnotte
        else:
            raise NoCagnotteRegistered(cagnotte_idx)
```

Reply:
But wait, shouldn't this be a part of SwagDB?

Reply:
I believe I'm right since this does not belong to the public SwagBank API

Reply:
LMAO why I didn't do that ?

Reply:
Mhhh IMO we should keep low-level function in db, like setter and getter. `get_active_cagnotte` is not a low-level function so I suggest to let it here

Reply:
Except this doesn't belong to the public API of `SwagBank` (which means, `SwagClient` won't and shouldn't ever use this function)
Therefore we have 3 choices:
1. Make this a regular function. I wouldn't recommand that, because it means decoupling it's meaning from the object it interacts with, and that doesn't make sense to me.
2. Make this a *private* method of `SwagBank`, because we juge this an implementation detail, but meaningful enough to warrant a function.
3. And the one I'd rather have, is to make this a *public* method of `SwagDB`. Why would we do that? Because `SwagDB` represent the data of `SwagBank`, and therefore the role of `SwagDB` is to provide every utility function needed to manipulate that data. To put it in other words methods of `SwagBank` should represent `Transaction`s operated on the blockchain in term of `SwagDB` operations. I'd also like to argue that "getting a `Cagnotte` we are actually allowed to touch" fits the bill for what I call a "low-level" (in `SwagDB`-levels) function, otherwise `Account` objects wouldn't be returned by `SwagDB` since they are an abstracted view over the tables of `SwagDB`.

Reply:
But wait, SwagClient use this function :V

like here https://github.com/BobbySoftwares/Bybbot/pull/7/files#diff-76a3e9d44095e56cab5f66a1e4d3760dc88f02887a680c2f90bb0f174d5a6965R567

and here https://github.com/BobbySoftwares/Bybbot/pull/7/files#diff-76a3e9d44095e56cab5f66a1e4d3760dc88f02887a680c2f90bb0f174d5a6965R528

Reply:
It need to use this fonction

Reply:
Well, After reading the bunch of code which use it, we can use the Cagnotte.info.currency. So nvm

Reply:
…aaaaaand this is a breach of API

Let's put aside the fact that in `donner` and `payer` we access the cagnotte two times. I don't see a better way so let's stay with that for this PR.

But never ever in existence should a view on `SwagDB` be accessed by `SwagClient` (even if, in Python it is impossible to forbid that). The whole point of having two classes, `Account` and `AccountInfo` is that one of them is a mutable view, while the other an immutable view. `SwagClient` should always, always be given the `*Info` version, because that's what is part of `SwagBank` public API.

Reply:
Well, let mettre get_active_cagnotte in the base de donné then